File: webmux/backend/src/api/ai.ts
/**
 * /api/ai — AI assistant for webmux terminal sessions.
 *
 * POST /api/ai/chat
 *   Body: { message: string, context?: string, sessionId?: string, history?: Message[] }
 *   Returns: { reply: string, model: string, ts: number }
 *
 * The request is forwarded to the RCC brain API (WEBMUX_RCC_URL + WEBMUX_RCC_TOKEN).
 * If RCC is not configured, falls back to the configured OPENAI_API_KEY or NVIDIA_API_KEY.
 *
 * Terminal context (last N lines of terminal output) is automatically prepended to the
 * system message so the assistant can explain errors or suggest next commands.
 */

import { Router, Request, Response } from 'express';

const router = Router();

interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

interface ChatRequest {
  message:   string;
  context?:  string;  // last N lines of terminal output
  sessionId?: string;
  history?:  Message[];
}

const SYSTEM_PROMPT = `You are a terminal assistant embedded in webmux, a browser-based terminal multiplexer.
You help users understand command output, diagnose errors, and suggest next steps.
When you see terminal output in <terminal_context> tags, use it to give specific, actionable advice.
Keep responses concise and focused. Prefer shell commands over lengthy explanations.
Format commands in backticks. Do not repeat the terminal context back to the user.`;

async function callRCCBrain(messages: Message[]): Promise<string> {
  const rccUrl   = process.env.WEBMUX_RCC_URL || process.env.LOOM_RCC_BRAIN_URL || '';
  const rccToken = process.env.WEBMUX_RCC_TOKEN || process.env.LOOM_RCC_AGENT_TOKEN || '';

  if (!rccUrl) throw new Error('RCC not configured');

  const resp = await fetch(`${rccUrl.replace(/\/$/, '')}/api/brain/request`, {
    method:  'POST',
    headers: {
      'Content-Type':  'application/json',
      'Authorization': `Bearer ${rccToken}`,
    },
    body: JSON.stringify({ messages, maxTokens: 512, priority: 'normal',
                           metadata: { source: 'webmux-ai' } }),
    signal: AbortSignal.timeout(15_000),
  });

  if (!resp.ok) throw new Error(`RCC HTTP ${resp.status}`);
  const data = await resp.json() as { status: string; result: string; error?: string };
  if (data.status !== 'completed') throw new Error(`Brain status: ${data.status}`);
  return data.result;
}

async function callDirectLLM(messages: Message[]): Promise<string> {
  const nvKey  = process.env.NVIDIA_API_KEY || '';
  const oaKey  = process.env.OPENAI_API_KEY || '';

  const apiKey = nvKey || oaKey;
  if (!apiKey) throw new Error('No LLM API key configured');

  const baseUrl = nvKey
    ? 'https://integrate.api.nvidia.com/v1'
    : 'https://api.openai.com/v1';
  const model = nvKey
    ? (process.env.WEBMUX_MODEL || 'meta/llama-3.1-70b-instruct')
    : (process.env.WEBMUX_MODEL || 'gpt-4o-mini');

  const resp = await fetch(`${baseUrl}/chat/completions`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${apiKey}` },
    body:    JSON.stringify({ model, messages, max_tokens: 512, temperature: 0.3 }),
    signal:  AbortSignal.timeout(20_000),
  });

  if (!resp.ok) throw new Error(`LLM API HTTP ${resp.status}`);
  const data = await resp.json() as {
    choices: Array<{ message: { content: string } }>;
    model: string;
  };
  return data.choices[0]?.message?.content?.trim() || '';
}

router.post('/chat', async (req: Request, res: Response) => {
  const { message, context, history = [] } = req.body as ChatRequest;

  if (!message || typeof message !== 'string' || message.trim().length === 0) {
    return res.status(400).json({ error: 'message required' });
  }

  // Build message array
  const messages: Message[] = [{ role: 'system', content: SYSTEM_PROMPT }];

  // Inject terminal context if provided
  const userContent = context && context.trim()
    ? `<terminal_context>\n${context.trim().slice(-3000)}\n</terminal_context>\n\n${message.trim()}`
    : message.trim();

  // Append conversation history (last 10 turns to keep context manageable)
  const recentHistory = history.slice(-10);
  messages.push(...recentHistory.filter(m => m.role !== 'system'));
  messages.push({ role: 'user', content: userContent });

  let reply = '';
  let model = 'unknown';
  let source = 'rcc';

  try {
    reply = await callRCCBrain(messages);
    model = 'rcc-brain';
  } catch (rccErr) {
    // Fallback to direct LLM
    source = 'direct';
    try {
      reply = await callDirectLLM(messages);
      model = process.env.WEBMUX_MODEL || 'llm';
    } catch (llmErr) {
      const errMsg = llmErr instanceof Error ? llmErr.message : String(llmErr);
      return res.status(503).json({
        error: 'AI assistant unavailable',
        detail: errMsg,
        hint: 'Set WEBMUX_RCC_URL+WEBMUX_RCC_TOKEN or NVIDIA_API_KEY/OPENAI_API_KEY',
      });
    }
  }

  return res.json({ reply, model, source, ts: Date.now() });
});

// GET /api/ai/status — check if AI is configured
router.get('/status', (_req: Request, res: Response) => {
  const hasRCC    = !!(process.env.WEBMUX_RCC_URL || process.env.LOOM_RCC_BRAIN_URL);
  const hasNVIDIA = !!process.env.NVIDIA_API_KEY;
  const hasOpenAI = !!process.env.OPENAI_API_KEY;
  const available = hasRCC || hasNVIDIA || hasOpenAI;
  res.json({
    available,
    providers: { rcc: hasRCC, nvidia: hasNVIDIA, openai: hasOpenAI },
    model: process.env.WEBMUX_MODEL || (hasNVIDIA ? 'meta/llama-3.1-70b-instruct' : 'gpt-4o-mini'),
  });
});

export default router;

File: webmux/server/internal/ai/ai.go
// Package ai forwards terminal chat to the configured RCC or direct provider.
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
	"unicode/utf16"

	"github.com/jordanhubbard/webmux/server/internal/config"
)

const SystemPrompt = `You are a terminal assistant embedded in webmux, a browser-based terminal multiplexer.
You help users understand command output, diagnose errors, and suggest next steps.
When you see terminal output in <terminal_context> tags, use it to give specific, actionable advice.
Keep responses concise and focused. Prefer shell commands over lengthy explanations.
Format commands in backticks. Do not repeat the terminal context back to the user.`
const Hint = "Set WEBMUX_RCC_URL+WEBMUX_RCC_TOKEN or NVIDIA_API_KEY/OPENAI_API_KEY"
const MaxResponseBytes = 1 << 20

type Message struct {
	Role       string `json:"role"`
	Content    string `json:"content"`
	rawContent json.RawMessage
}

func (m Message) MarshalJSON() ([]byte, error) {
	if m.rawContent != nil {
		return json.Marshal(struct {
			Role    string          `json:"role"`
			Content json.RawMessage `json:"content"`
		}{m.Role, m.rawContent})
	}
	type plain Message
	return json.Marshal(plain(m))
}

type Request struct {
	Message string    `json:"message"`
	Context string    `json:"context"`
	History []Message `json:"history"`
}
type Response struct {
	Reply     string `json:"reply"`
	Model     string `json:"model"`
	Source    string `json:"source"`
	Timestamp int64  `json:"ts"`
}
type Providers struct {
	RCC    bool `json:"rcc"`
	NVIDIA bool `json:"nvidia"`
	OpenAI bool `json:"openai"`
}
type Status struct {
	Available bool      `json:"available"`
	Providers Providers `json:"providers"`
	Model     string    `json:"model"`
}
type Service struct {
	env    func(string) string
	client *http.Client
}

func New(env func(string) string, client *http.Client) *Service {
	if env == nil {
		env = os.Getenv
	}
	if client == nil {
		client = &http.Client{CheckRedirect: func(next *http.Request, previous []*http.Request) error {
			if len(previous) >= 10 {
				return errors.New("provider redirect limit exceeded")
			}
			first := previous[0].URL
			if !strings.EqualFold(first.Scheme, next.URL.Scheme) || !strings.EqualFold(first.Host, next.URL.Host) {
				return http.ErrUseLastResponse
			}
			return nil
		}}
	}
	return &Service{env: env, client: client}
}
func first(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}
func (s *Service) rccURL() string { return first(s.env("WEBMUX_RCC_URL"), s.env("LOOM_RCC_BRAIN_URL")) }
func (s *Service) Status() Status {
	p := Providers{RCC: s.rccURL() != "", NVIDIA: s.env("NVIDIA_API_KEY") != "", OpenAI: s.env("OPENAI_API_KEY") != ""}
	model := "gpt-4o-mini"
	if p.NVIDIA {
		model = "meta/llama-3.1-70b-instruct"
	}
	return Status{Available: p.RCC || p.NVIDIA || p.OpenAI, Providers: p, Model: first(s.env("WEBMUX_MODEL"), model)}
}

func messages(input Request) []Message {
	result := []Message{{Role: "system", Content: SystemPrompt}}
	history := input.History
	if len(history) > 10 {
		history = history[len(history)-10:]
	}
	for _, message := range history {
		if message.Role != "system" {
			result = append(result, message)
		}
	}
	user := Message{Role: "user", Content: config.TrimSpace(input.Message)}
	if terminal := config.TrimSpace(input.Context); terminal != "" {
		units := utf16.Encode([]rune(terminal))
		if len(units) > 3000 {
			units = units[len(units)-3000:]
		}
		prefix, suffix := "<terminal_context>\n", "\n</terminal_context>\n\n"+user.Content
		user.Content = prefix + string(utf16.Decode(units)) + suffix
		// JavaScript slice counts UTF-16 units and can leave a leading low
		// surrogate. Preserve that escaped unit instead of replacing it with U+FFFD.
		if units[0] >= 0xdc00 && units[0] <= 0xdfff {
			a, _ := json.Marshal(prefix)
			b, _ := json.Marshal(string(utf16.Decode(units[1:])) + suffix)
			user.rawContent = []byte(string(a[:len(a)-1]) + fmt.Sprintf(`\u%04x`, units[0]) + string(b[1:]))
		}
	}
	return append(result, user)
}

func (s *Service) post(ctx context.Context, url, key string, payload any, timeout time.Duration, provider string, out any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return errors.New("Invalid provider request")
	}
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(data))
	if err != nil {
		return errors.New("Invalid provider URL")
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+key)
	response, err := s.client.Do(req)
	if err != nil {
		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
			return errors.New("The operation was aborted due to timeout")
		}
		return errors.New("fetch failed")
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("%s HTTP %d", provider, response.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(response.Body, MaxResponseBytes+1))
	if err != nil {
		return errors.New("Invalid provider response")
	}
	if len(body) > MaxResponseBytes {
		return errors.New("Provider response too large")
	}
	if err := json.Unmarshal(body, out); err != nil {
		return errors.New("Invalid provider response")
	}
	return nil
}

func (s *Service) rcc(ctx context.Context, messages []Message) (string, error) {
	url := s.rccURL()
	if url == "" {
		return "", errors.New("RCC not configured")
	}
	var response struct {
		Status string  `json:"status"`
		Result *string `json:"result"`
	}
	err := s.post(ctx, strings.TrimSuffix(url, "/")+"/api/brain/request", first(s.env("WEBMUX_RCC_TOKEN"), s.env("LOOM_RCC_AGENT_TOKEN")), map[string]any{
		"messages": messages, "maxTokens": 512, "priority": "normal", "metadata": map[string]string{"source": "webmux-ai"},
	}, 15*time.Second, "RCC", &response)
	if err != nil {
		return "", err
	}
	if response.Status != "completed" {
		return "", fmt.Errorf("Brain status: %s", response.Status)
	}
	if response.Result == nil {
		return "", errors.New("Invalid provider response")
	}
	return *response.Result, nil
}

func (s *Service) direct(ctx context.Context, messages []Message) (string, error) {
	nvKey, oaKey := s.env("NVIDIA_API_KEY"), s.env("OPENAI_API_KEY")
	key := first(nvKey, oaKey)
	if key == "" {
		return "", errors.New("No LLM API key configured")
	}
	url, model := "https://api.openai.com/v1/chat/completions", "gpt-4o-mini"
	if nvKey != "" {
		url, model = "https://integrate.api.nvidia.com/v1/chat/completions", "meta/llama-3.1-70b-instruct"
	}
	var response struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	err := s.post(ctx, url, key, map[string]any{"model": first(s.env("WEBMUX_MODEL"), model), "messages": messages, "max_tokens": 512, "temperature": 0.3}, 20*time.Second, "LLM API", &response)
	if err != nil {
		return "", err
	}
	if response.Choices == nil {
		return "", errors.New("Invalid provider response")
	}
	if len(response.Choices) == 0 {
		return "", nil
	}
	return config.TrimSpace(response.Choices[0].Message.Content), nil
}

func (s *Service) Chat(ctx context.Context, input Request) (Response, error) {
	messages := messages(input)
	reply, err := s.rcc(ctx, messages)
	model, source := "rcc-brain", "rcc"
	if err != nil {
		if ctx.Err() != nil {
			return Response{}, ctx.Err()
		}
		reply, err = s.direct(ctx, messages)
		if err != nil {
			return Response{}, err
		}
		model, source = first(s.env("WEBMUX_MODEL"), "llm"), "direct"
	}
	return Response{Reply: reply, Model: model, Source: source, Timestamp: time.Now().UnixMilli()}, nil
}
